File: sdks/python/src/echo_agent_sdk/memory_scope_values.py
from __future__ import annotations

from enum import Enum


class MemoryScope(str, Enum):
    """Memory lifetime scopes with Rust-compatible wire names and priority."""

    USER = "user"
    PROJECT = "project"
    REPO = "repo"
    TASK = "task"
    SESSION = "session"
    RUN = "run"


_MEMORY_SCOPE_ALL: tuple[MemoryScope, ...] = (
    MemoryScope.USER,
    MemoryScope.PROJECT,
    MemoryScope.REPO,
    MemoryScope.TASK,
    MemoryScope.SESSION,
    MemoryScope.RUN,
)


def memory_scope_all() -> tuple[MemoryScope, ...]:
    return _MEMORY_SCOPE_ALL


def memory_scope_name(scope: MemoryScope) -> str:
    scope = _validate(scope)
    return scope.value


def memory_scope_priority(scope: MemoryScope) -> int:
    scope = _validate(scope)
    return _MEMORY_SCOPE_ALL.index(scope)


def memory_scope_is_persistent(scope: MemoryScope) -> bool:
    scope = _validate(scope)
    return scope in {MemoryScope.USER, MemoryScope.PROJECT, MemoryScope.REPO}


def parse_memory_scope(value: str) -> MemoryScope | None:
    if not isinstance(value, str):
        return None
    return {
        "user": MemoryScope.USER,
        "project": MemoryScope.PROJECT,
        "proj": MemoryScope.PROJECT,
        "repo": MemoryScope.REPO,
        "task": MemoryScope.TASK,
        "session": MemoryScope.SESSION,
        "sess": MemoryScope.SESSION,
        "run": MemoryScope.RUN,
    }.get(value.strip().lower())


def _validate(scope: MemoryScope) -> MemoryScope:
    if not isinstance(scope, MemoryScope):
        raise TypeError(f"unknown memory scope: {scope!r}")
    return scope

File: sdks/typescript/src/memory_scope_values.ts
/** Memory lifetime scopes with Rust-compatible wire names and priority. */
export const MemoryScope = Object.freeze({
  User: "user",
  Project: "project",
  Repo: "repo",
  Task: "task",
  Session: "session",
  Run: "run",
} as const);

export type MemoryScope = (typeof MemoryScope)[keyof typeof MemoryScope];

const MEMORY_SCOPE_ALL: readonly MemoryScope[] = Object.freeze([
  MemoryScope.User,
  MemoryScope.Project,
  MemoryScope.Repo,
  MemoryScope.Task,
  MemoryScope.Session,
  MemoryScope.Run,
]);
const MEMORY_SCOPES: ReadonlySet<string> = new Set(MEMORY_SCOPE_ALL);

export function memoryScopeAll(): readonly MemoryScope[] {
  return MEMORY_SCOPE_ALL;
}

export function memoryScopeName(scope: MemoryScope): string {
  validateMemoryScope(scope);
  return scope;
}

export function memoryScopePriority(scope: MemoryScope): number {
  validateMemoryScope(scope);
  return MEMORY_SCOPE_ALL.indexOf(scope);
}

export function memoryScopeIsPersistent(scope: MemoryScope): boolean {
  validateMemoryScope(scope);
  return scope === MemoryScope.User || scope === MemoryScope.Project || scope === MemoryScope.Repo;
}

export function parseMemoryScope(value: string): MemoryScope | null {
  if (typeof value !== "string") return null;
  switch (value.trim().toLowerCase()) {
    case "user": return MemoryScope.User;
    case "project": case "proj": return MemoryScope.Project;
    case "repo": return MemoryScope.Repo;
    case "task": return MemoryScope.Task;
    case "session": case "sess": return MemoryScope.Session;
    case "run": return MemoryScope.Run;
    default: return null;
  }
}

function validateMemoryScope(scope: unknown): asserts scope is MemoryScope {
  if (typeof scope !== "string" || !MEMORY_SCOPES.has(scope)) {
    throw new TypeError(`unknown memory scope: ${String(scope)}`);
  }
}
